core/lib/zksync_core/src/metadata_calculator/recovery/tests.rs

```diff
@@ -0,0 +1,431 @@
+        RecoveryWorkflowCase::Stop => {
+            let tree_info = tree_reader.await.info().await;
+            assert_eq!(tree_info.root_hash, snapshot_recovery.l1_batch_root_hash);
+            assert_eq!(tree_info.leaf_count, 200);
+            assert_eq!(
+                tree_info.next_l1_batch_number,
+                snapshot_recovery.l1_batch_number + 1
+            );
+        }
+
+        // Emulate state keeper adding a new L1 batch to Postgres.
+        RecoveryWorkflowCase::CreateBatch => {
+            tree_reader.await;
+
+            let mut storage = storage.start_transaction().await.unwrap();
+            let mut new_logs = gen_storage_logs(500..600, 1).pop().unwrap();
+            // Logs must be sorted by `log.key` to match their enum index assignment
+            new_logs.sort_unstable_by_key(|log| log.key);
+
+            extend_db_state_from_l1_batch(
+                &mut storage,
+                snapshot_recovery.l1_batch_number + 1,
+                [new_logs.clone()],
+            )
+            .await;
+            storage.commit().await.unwrap();
+
+            // Wait until the inserted L1 batch is processed by the calculator.
+            let new_root_hash = loop {
+                let (next_l1_batch, root_hash) = delay_rx.recv().await.unwrap();
+                if next_l1_batch == snapshot_recovery.l1_batch_number + 2 {
+                    break root_hash;
+                }
+            };
+
+            let all_tree_instructions: Vec<_> = snapshot_logs
+                .iter()
+                .chain(&new_logs)
+                .enumerate()
+                .map(|(i, log)| TreeInstruction::write(log.key, i as u64 + 1, log.value))
+                .collect();
+            let expected_new_root_hash =
+                ZkSyncTree::process_genesis_batch(&all_tree_instructions).root_hash;
+            assert_ne!(expected_new_root_hash, snapshot_recovery.l1_batch_root_hash);
+            assert_eq!(new_root_hash, expected_new_root_hash);
+        }
+    }
+
+    stop_sender.send_replace(true);
+    calculator_task.await.expect("calculator panicked").unwrap();
+}
+
+/// Prepares a recovery snapshot without performing genesis.
+async fn prepare_clean_recovery_snapshot(
+    storage: &mut StorageProcessor<'_>,
+    snapshot_logs: &[StorageLog],
+) -> SnapshotRecoveryStatus {
+    let written_keys: Vec<_> = snapshot_logs.iter().map(|log| log.key).collect();
+    let tree_instructions: Vec<_> = snapshot_logs
+        .iter()
+        .enumerate()
+        .map(|(i, log)| TreeInstruction::write(log.key, i as u64 + 1, log.value))
+        .collect();
+    let l1_batch_root_hash = ZkSyncTree::process_genesis_batch(&tree_instructions).root_hash;
+
+    storage
+        .protocol_versions_dal()
+        .save_protocol_version_with_tx(ProtocolVersion::default())
+        .await;
+    // TODO (PLA-596): Don't insert L1 batches / miniblocks once the relevant foreign keys are removed
```

Comment:
Because of miniblock / L1 batch-related foreign key constraints, it's impossible to precisely recreate Postgres state after it's recovered from a snapshot right now. Correspondingly, the new tests don't *precisely* test the full recovery life cycle because of this. I expect that these constraints will be removed in the referenced task, and then this code could be removed as well.